handler/src/main/java/io/netty/handler/ssl/SniHandler.java

```diff
@@ -313,31 +333,58 @@ public void operationComplete(Future<SslContext> future) throws Exception {
     }
 
     /**
-     * The default implementation will simply call {@link AsyncMapping#map(Object, Promise)} but
-     * users can override this method to implement custom behavior.
+     * The default implementation will simply call {@link SslContextMapping#map(String, Promise)} but users can
+     * override this method to implement custom behavior.
      *
-     * @see AsyncMapping#map(Object, Promise)
+     * NOTE: This method returns a {@code Future<?>} to accommodate the old and new way of doing things without
+     * braking the API. Users who override this method are advised to change their return value from
+     * {@code Future<SslContext>} to {@code Future<Supplier<? extends SslContext>>}.
+     *
+     * @see SslContextMapping#map(String, Promise)
+     * @deprecated the return value will at some point change from {@code Future<SslContext>} to
```

Comment:
There is unfortunately a little problem with that. What does SniHandler's `select(...)` method call? If it's `lookup(...)` then the whole Supplier thing gets rendered useless because we do want `replaceHandler(..., Supplier)` to be called.

If it's `lookupSupplier(...)` then here's no way to accommodate the `lookup(...)` into it.